spyder/api/plugins.py

```diff
@@ -528,8 +796,159 @@ def toggle_view(self, checked):
         else:
             self.dockwidget.hide()
 
+    # FIXME: What are some examples of this usage?
     def set_ancestor(self, ancestor):
         """
         Needed to update the ancestor/parent of child widgets when undocking.
         """
         pass
+
+    # --- API: available methods ---------------------------------------------
+    # ------------------------------------------------------------------------
+    # FIXME: This SHOULD be private and called from mainwindow.py so users
+    # do not need to know about this on plugin registration
+    def add_dockwidget(self):
+        """Add the plugin's QDockWidget to the main window."""
+        self.main.add_dockwidget(self)
+
+        # This is not necessary for the Editor because it calls
+        # _setup directly on init.
+        if self.CONF_SECTION != 'editor':
+            self._setup()
+
+    # TODO: Update to use name of plugin instead of plugin instance
+    #########################################################################
+    # FIXME: If we turn this into a class attribute, users do not need to
+    # take care of this logic, but only define next to which plugin (by name)
+    # they want to dock their plugin.
+    def tabify(self, core_plugin):
+        """
+        Tabify plugin next to one of the core plugins.
+
+        Parameters
+        ----------
+        core_plugin: SpyderPluginWidget
+            Core Spyder plugin this one will be tabified next to.
+
+        Examples
+        --------
+        >>> self.tabify(self.main.variableexplorer)
+        >>> self.tabify(self.main.ipyconsole)
+
+        Notes
+        -----
+        The names of variables associated with each of the core plugins
+        can be found in the `setup` method of `MainWindow`, present in
+        `spyder/app/mainwindow.py`.
+        """
+        self.main.tabify_plugins(core_plugin, self)
+
+    # TODO: Rethink this method
+    # A widget should expose its actions in a reliable preditcable way
+    # and the bindning to shortcuts should be performed at the plugin
+    # level only
+    def register_widget_shortcuts(self, widget):
+        """
+        Register shortcuts defined by a plugin's widget so they take
+        effect when the plugin is focused.
+
+        Parameters
+        ----------
+        widget: QWidget
+            Widget to register shortcuts for.
+
+        Notes
+        -----
+        The widget interface must have a method called
+        `get_shortcut_data` for this to work. Please see
+        `spyder/widgets/findreplace.py` for an example.
+        """
+        for qshortcut, context, name in widget.get_shortcut_data():
+            self.register_shortcut(qshortcut, context, name)
+
+    # FIXME: This should probably just emit a signal and handle this on
+    # the mainwindow.py directly. Calls to self.main inside a plugin are
+    # Funky!
```

Comment:
Please leave this method as is or change it in another PR. It took me a hell of lot of time and effort to correctly define the logic for this and check every action and button to verify that it was working correctly.

Reply:
Not an option anymore.